llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp

```diff
@@ -0,0 +1,116 @@
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/DerivedTypes.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+#include "llvm/Transforms/Utils/ValueMapper.h"
+#include <map>
+#include <vector>
+
+namespace {
+llvm::BasicBlock *splitBlockBefore(llvm::BasicBlock *Old,
+                                   llvm::Instruction *SplitPt,
+                                   llvm::DomTreeUpdater *DTU = nullptr,
+                                   llvm::LoopInfo *LI = nullptr,
+                                   llvm::MemorySSAUpdater *MSSAU = nullptr,
+                                   const llvm::Twine &BBName = "",
+                                   bool Before = true) {
+  return llvm::SplitBlock(Old, SplitPt, DTU, LI, MSSAU, BBName, Before);
+}
+
+class KulaginMagicInlinePass
+    : public llvm::PassInfoMixin<KulaginMagicInlinePass> {
+public:
+  llvm::PreservedAnalyses run(llvm::Function &Func,
+                              llvm::FunctionAnalysisManager &) {
+    std::vector<llvm::Instruction *> InstrDeleteList;
+    llvm::LLVMContext &CTX = Func.getContext();
+    for (llvm::BasicBlock &Block : Func) {
+      for (llvm::Instruction &Instr : Block) {
+        if (llvm::CallInst *CallInstruction =
+                llvm::dyn_cast<llvm::CallInst>(&Instr)) {
+          llvm::Function *Called = CallInstruction->getCalledFunction();
+          if (Called == &Func || !Called->getReturnType()->isVoidTy() ||
+              Called->getFunctionType()->getNumParams() > 0) {
+            continue;
+          }
+          std::map<llvm::BasicBlock *, llvm::BasicBlock *> BlockMap;
+          llvm::ValueToValueMapTy VMap;
+          llvm::BasicBlock *SplittedBlock = splitBlockBefore(&Block, &Instr);
+          llvm::BasicBlock *NextBlockPlace = SplittedBlock->getNextNode();
+          for (llvm::BasicBlock &ToInlineBasicBlock : *Called) {
+            llvm::BasicBlock *TmpBlock =
+                llvm::BasicBlock::Create(CTX, "", &Func, NextBlockPlace);
+            BlockMap[&ToInlineBasicBlock] = TmpBlock;
+            NextBlockPlace = TmpBlock->getNextNode();
+          }
+          for (llvm::BasicBlock &ToInlineBasicBlock : *Called) {
```

Comment:
May these loops be joint? I mean we split a block in a processed function and loop through blocks and instructions of inlined functions inserting them in place?

Reply:
> May these loops be joint? I mean we split a block in a processed function and loop through blocks and instructions of inlined functions inserting them in place?

No. Because some branch instructions can call a block that was not created and associated yet. Basically any loop and if statement. The branch instruction there calls the next block.

Reply:
Ok, good.